CIP-0030/README.md

```diff
@@ -0,0 +1,250 @@
+
+
+# Motivation
+
+In order to facilitate future dApp development, we will need a way for dApps to communicate with the user's wallet. While Cardano does not yet support smart contracts, there are still various use cases for this, such as NFT management. This will also lay the groundwork for an updated version of the spec once the Alonzo hardfork is released which can extend it to allow for Plutus support.
+
+
+
+# Specification
+
+## Version
+
+The API specified in this document will count as version 0.1.0 for version-checking purposes below.
+
+## Data Types
+
+### Bytes
+
+A hex-encoded string of the corresponding bytes.
+
+### cbor\<T>
+
+A hex-encoded string representing [CBOR](https://tools.ietf.org/html/rfc7049) corresponding to `T` defined via [CDDL](https://tools.ietf.org/html/rfc8610) either inside of the [Shelley Mult-asset binary spec](https://github.com/input-output-hk/cardano-ledger-specs/blob/0738804155245062f05e2f355fadd1d16f04cd56/shelley-ma/shelley-ma-test/cddl-files/shelley-ma.cddl) or, if not present there, from the [CIP-0008 signing spec](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0008/CIP-0008.md).
+This representation was chosen when possible as it is consistent across the Cardano ecosystem and widely used by other tools, such as [cardano-serialization-lib](https://github.com/Emurgo/cardano-serialization-lib), which has support to encode every type in the binary spec as CBOR bytes.
+
+### TransactionUnspentOutput
+
+If we have CBOR specified by the following CDDL referencing the Shelley-MA CDDL:
+```cddl
+transaction_unspent_output = [
+  input: transaction_input,
+  output: transaction_output,
+]
+```
+then we define
+```
+type TransactionUnspentOutput = cbor<transaction_unspent_output>
+```
+
+This allows us to use the output for constructing new transactions using it as an output as the `transaction_output` in the Shelley Multi-asset CDDL does not contain enough information on its own to spend it.
+
+### Paginate
+
+```
+type Paginate = {|
+  page: number,
+  limit: number,
+|};
+```
+Used to specify optional pagination for some API calls. Limits results to {limit} each page, and uses a 0-indexing {page} to refer to which of those pages of {limit} items each. dApps should be aware that if a wallet is modified between paginated calls that this will change the pagination, e.g. some results skipped or showing up multiple times but otherwise the wallet must respect the pagination order.
+
+
+## Error Types
+
+### APIError
+
+```
+APIErrorCode {
+	InvalidRequest: -1,
+	InternalError: -2,
+	Refused: -3,
+	AccountChange: 4,
+}
+APIError {
+	code: APIErrorCode,
+	info: string
+}
+```
```

Comment:
Currently implementing this.

Was the - before 1,2,3 intended?

Reply:
@MarcelKlammer  Yes it's intended so that they are distinct from the non-negative error codes specific to certain calls. However, that should be -4. I just fixed that in a final commit just now.